Compile error by using reference of SumType as struct field

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.3.1 d7758b2.69c9d47
**OS:**  macos, macOS, 12.6, 21G115

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
struct S1 {}

struct S2 {}

type T1 = S1 | S2

struct SumTypePtr {
	ptr &T1
}

fn main() {
	val := S1{}
	s := SumTypePtr{
		ptr: &val
	}
	println(s)
}
```

**What did you expect to see?**
compile without error.

I'm not sure the following is appropriate, but can be something like following.
```
SumTypePtr{
    ptr: &T1
}
```
 
**What did you see instead?**
```
sumtype_ptr_str.v 
==================
/tmp/v_502/sumtype_ptr_str.11254321061033638332.tmp.c:2052:46: error: indirection requires pointer operand ('string' (aka 'struct string') invalid)
        string _t1 = isnil(it.ptr) ? _SLIT("nil") : *indent_main__T1_str(*it.ptr, indent_count + 1);
                                                    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_502/sumtype_ptr_str.11254321061033638332.tmp.c:2209:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
1 warning and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

--- a/vlib/v/gen/c/auto_str_methods.v
+++ b/vlib/v/gen/c/auto_str_methods.v
@@ -921,8 +921,8 @@ fn (mut g Gen) gen_str_for_struct(info ast.Struct, styp string, str_fn_name stri
 			funcprefix += 'isnil(it.${c_name(field.name)})'
 			funcprefix += ' ? _SLIT("nil") : '
 			// struct, floats and ints have a special case through the _str function
-			if sym.kind !in [.struct_, .alias, .enum_, .interface_] && !field.typ.is_int_valptr()
-				&& !field.typ.is_float_valptr() {
+			if sym.kind !in [.struct_, .alias, .enum_, .sum_type, .map, .interface_]
+				&& !field.typ.is_int_valptr() && !field.typ.is_float_valptr() {
 				funcprefix += '*'
 			}
 		}

--- a/vlib/v/gen/c/testdata/ref_sumtype_as_struct_field.out
+++ b/vlib/v/gen/c/testdata/ref_sumtype_as_struct_field.out
@@ -1,0 +1,3 @@
+SumTypePtr{
+    ptr: &nil
+}

--- a/vlib/v/gen/c/testdata/ref_sumtype_as_struct_field.vv
+++ b/vlib/v/gen/c/testdata/ref_sumtype_as_struct_field.vv
@@ -1,0 +1,14 @@
+struct S1 {}
+
+struct S2 {}
+
+type T1 = S1 | S2
+
+struct SumTypePtr {
+	ptr &T1 = unsafe { nil }
+}
+
+fn main() {
+	ptr := SumTypePtr{}
+	println(ptr)
+}

